### PR TITLE
Tidying prarameter for the map matching plugin.

### PR DESCRIPTION
**Pull request by gardster**
# Issue

#3778

Adds special parameter to turn off track splitting. I think this parameter will be useful also in other tidying functions, like in #3149 
Default value does not change current map matching behavior.

## Tasklist
 - [x] update relevant [Wiki pages](https://github.com/Project-OSRM/osrm-backend/wiki)
 - [x] review
 - [x] adjust for comments

### CODE REVIEW

**Comment by daniel-j-h** [comment]
Could you add a changelog entry. Travis is failing because of clang-format check.


Also there is a trace tidying implementation for map matching [here](https://github.com/Project-OSRM/osrm-backend/pull/3149) which needs to be hooked up to the plugin. It does exactly the opposite of what you are doing here: it removes too many trace points from the trace for blob removal. Not sure if we should call this parameter `tidying` and have it on by default. Can we unify both PRs?

**Comment by gardster** [comment]
Ok. I cherry-picked #3149 functions and added it to the matching logic. So now this parameter called `preprocess` and has 3 options: 

- `false` - without any preprocessing, or split identification;
- `simple` - current matching behavior, when the plugin checks timestamps gaps. I think that this option is good by default, because it keeps current behavior;
- `full` option - for #3149 logic.

At what release note I need to add a changelog entry?

**Comment by danpat** [comment]
@gardster Just put it in the top section - we'll merge this into master, then we usually edit the CHANGELOG when it comes time to branch the next release.

**Comment by gardster** [comment]
@danpat I added a changelog entry.

#3149 as is fail the CI tests. Maybe we ommit this functionality from current PR (but keep all flag options), and include track tidying when it will be ready?

**Comment by TheMarex** [comment]
API wise I don't think `tidy` is the right word for what you are trying to disable. Nor is this really connected to @daniel-j-h PR, but thanks for integrating it anyway! Proposal:

- `gaps=split,ignore` (`ignore` enables the functionality you implemented)
- `tidy=true,false` (implements what @daniel-j-h has implemented)

However the build is failing to compile right now, so I'll mark this as WIP.

**Comment by gardster** [comment]
@TheMarex @daniel-j-h  I split options to `gaps` and `tidy`, fix build failure and add some additional logic to handle tidied coordinates. Please, take another look to this PR.

File: docs/http.md [hunk 1]
```diff
@@ -279,6 +279,8 @@ In addition to the [general options](#general-options) the following options are
 |overview    |`simplified` (default), `full`, `false`         |Add overview geometry either full, simplified according to highest zoom level it could be display on, or not at all.|
 |timestamps  |`{timestamp};{timestamp}[;{timestamp} ...]`     |Timestamps for the input locations in seconds since UNIX epoch. Timestamps need to be monotonically increasing. |
 |radiuses    |`{radius};{radius}[;{radius} ...]`              |Standard deviation of GPS precision used for map matching. If applicable use GPS accuracy.|
+|gaps        |`split` (default), `ignore`                     |Allows the input track splitting based on huge timestamp gaps between points.             |
+|tidying     |`true`, `false` (default)                       |Allows the input track modification to obtain better matching quality for noisy tracks.   |
```
Comment (TheMarex) [inline docs/http.md]: Can you rename the option to `tidy` (verb)? Thanks!

File: include/engine/api/match_parameters.hpp [hunk 2]
```diff
@@ -50,23 +50,37 @@ namespace api
...
+                    GapsType gaps_processing_,
+                    bool use_tidying_,
+                    Args... args_)
+        : RouteParameters{std::forward<Args>(args_)...}, timestamps{std::move(timestamps_)},
+          gaps_processing(gaps_processing_), use_tidying(use_tidying_)
     {
     }
 
     std::vector<unsigned> timestamps;
+    GapsType gaps_processing;
```
Comment (TheMarex) [inline include/engine/api/match_parameters.hpp]: Just for consistency I would try to name these like the query option: `gaps` and `tidy`.

**Comment by gardster** [comment]
Fixed

**Comment by TheMarex** [comment]
@gardster thanks for getting this in. 👍

**Pull request merged**